vignettes/usage-and-migration.Rmd

```diff
@@ -0,0 +1,340 @@
+
+We recommend adding a build to your matrix that targets the devel version of tidyr:
+
+* Always? If your package is tightly coupled to tidyr, consider leaving this 
+  build in place all the time, so you know if changes in tidyr affect your 
+  package. Modifiers like `allow_failures` and `fast_finish` may be useful. 
+  Read more about the [Travis build matrix](https://docs.travis-ci.com/user/customizing-the-build/#build-matrix).
+  
+* Right before a tidyr release? For everyone else, you could just add a 
+  tidyr-devel build during the period preceding a major tidyr release that has 
+  the potential for breaking changes.
+
+Example of `.travis.yml` config that tests against R devel, release, and oldrel, assesses test coverage, and includes a build against devel tidyr:
+
+``` yaml
+# R for travis: see documentation at https://docs.travis-ci.com/user/languages/r
+
+language: R
+cache: packages
+
+matrix:
+  include:
+  - r: devel
+  - r: release
+    after_success:
+    - Rscript -e 'covr::codecov()'
+  - r: release
+    name: tidyr-devel
+    before_script: Rscript -e "remotes::install_github('tidyverse/tidyr')"
+  - r: oldrel
+```
+
+## tidyr 0.8.3 -> 1.0.0 migration
+
+tidyr 1.0.0 (tidyr 0.8.99.9000 in the development version) makes considerable changes to the interface of `nest()` and `unnest()` in order to bring them inline with newer tidyverse conventions. I have tried to make the functions as backward compatible as possible and to give informative warning messages, but I could not cover 100% of use cases, so you may need to change your package code.
+
+Ideally, you'll tweak your package so that it works with both tidyr 0.8.3 and tidyr 1.0.0. This makes life considerably easier because it means there's no need to coordinate CRAN submissions - you can submit your package that works with both tidyr versions, before I submit tidyr to CRAN. This section describes our recommend practices for doing so, drawing from the general principles described in <https://principles.tidyverse.org/changes-multivers.html>.
+
+If you use travis already, we **strongly** recommend adding a build that tests with the development version of tidyr; see above for details.
+
+### Conditional code
+
+Sometimes you'll be able to write code that works with v0.8.3 and v1.0.0. But this often forces you to write code that's not particularly natural for either version and you'd be better off to (temporarily) have separate code paths, each containing non-contrived code. You get to re-use your existing code in the "old" branch, which will eventually be phased out, and write clean, forward-looking code in the "new" branch.
+
+The basic template looks like this:
+
+```{r}
+tidyr_new_interface <- function() {
+  packageVersion("tidyr") > "0.8.99"
```

Comment:
Should we mention the importance of doing this in a function? I.e. so that it's determined at run time not install time? Otherwise, I think people could easily get creative and "simplify" this.